src/rpc/pd/leader.rs

```diff
@@ -0,0 +1,432 @@
+// Copyright 2018 The TiKV Project Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    sync::{Arc, RwLock},
+    thread::{self, JoinHandle},
+    time::{Duration, Instant},
+};
+
+use futures::{
+    sync::{
+        mpsc::{unbounded, UnboundedReceiver, UnboundedSender},
+        oneshot,
+    },
+    Future, Sink, Stream,
+};
+use fxhash::FxHashSet as HashSet;
+use grpcio::{CallOption, Environment, WriteFlags};
+use kvproto::{pdpb, pdpb_grpc};
+use log::*;
+use tokio_core::reactor::{Core, Handle as OtherHandle};
+
+use crate::{
+    rpc::{
+        pd::{
+            context::{observe_tso_batch, request_context},
+            PdTimestamp,
+        },
+        security::SecurityManager,
+        util::HandyRwLock,
+    },
+    Error, Result,
+};
+
+macro_rules! pd_request {
+    ($cluster_id:expr, $type:ty) => {{
+        let mut request = <$type>::new();
+        let mut header = ::kvproto::pdpb::RequestHeader::new();
+        header.set_cluster_id($cluster_id);
+        request.set_header(header);
+        request
+    }};
+}
+
+type TsoChannel = oneshot::Sender<PdTimestamp>;
+
+enum PdTask {
+    Init,
+    Request,
+    Response(Vec<oneshot::Sender<PdTimestamp>>, pdpb::TsoResponse),
+}
+
+struct PdReactor {
+    task_tx: Option<UnboundedSender<Option<PdTask>>>,
```

Comment:
Unbounded always gives me the shivers. 🙈

Reply:
Let me investigate it one more time. Since only TSO requests and responses are dispatched by reactor thread now. We might be able to change it to bounded this time.